### PR TITLE
fix(melange): depend on selected impl when setting up JS rules for virtual lib

### LINKED ISSUE #7104: Inconsistent behaviour for virtual libraries using melange mode
## Expected Behavior

When creating a virtual library in native or byte modes, it is possible to assign a module to be virtual and add an implementation and interface that wraps around the virtual module in the virtual library.

Specifically the following structure illustrates this:

```
- vlib/
  - virt.mli     # virtual module
  - vlib.mli     # interface of vlib
  - vlib.ml      # implementation of vlib (uses virt!)
- vlib_impl_1/
  - virt.ml      # implementation of the virt module for vlib
```

This works as expected in native and byte modes and allows providing multiple implementations for `virt.mli` while offering a high-level interface in `vlib.mli`.

This doesn't work in melange mode.

## Actual Behavior

This doesn't work when `(modes melange)` is used. Not only it is not possible to replicate the described workflow, but having `(modes melange)` also prevents the non-melange implementation to be compiled.

Error when compiling a melange target:
```
$ dune build @melange
File "vlib/vlib.ml", line 1:        
Error: Virt not found, it means either the module does not exist or it is a namespace
```

Additionally, if `(modes melange)` is present in `vlib`, the native compilation fails with:
```
File "src/dune_rules/virtual_rules.ml", line 11, characters 30-37:
Error: No rule found for vlib/.vlib.objs/byte/virt.cmi
File "src/dune_rules/virtual_rules.ml", line 11, characters 30-37:
Error: No rule found for vlib/.vlib.objs/byte/vlib.cmi
File "src/dune_rules/virtual_rules.ml", line 11, characters 30-37:
Error: No rule found for vlib/.vlib.objs/byte/vlib.cmo
File "src/dune_rules/virtual_rules.ml", line 11, characters 30-37:
Error: No rule found for vlib/.vlib.objs/native/vlib.cmx
File "src/dune_rules/virtual_rules.ml", line 11, characters 30-37:
Error: No rule found for vlib/.vlib.objs/native/vlib.o
```

## Reproduction

The following repository fully replicates the above scenario based on the [blackbox test](https://github.com/ocaml/dune/tree/main/test/blackbox-tests/test-cases/melange/virtual_lib.t) included in dune:

- https://github.com/rizo/dune-melange-virt-issue

## Specifications

- Version of `dune`:  ~3.7 `05797cd3f505beca4d6d62c576506d7608a6d29d`
- Version of `melange`: `f4ddebbc7755af5e983c4df2ac51498f286e1074`
- Version of `ocaml`: 4.14.0
- Operating system (distribution and version): 22.3.0 Darwin

### PR DESCRIPTION
**Pull request by anmonteiro**
fixes #7104

TODO:

- this change depends on https://github.com/melange-re/melange/pull/1067. need to:
    - [x] port it back to the melange 4.14 branch so that we can run it in dune's tests
    - [x] depend on the melange development version for CI in dune

Even though the PR is in draft, I'm requesting your reviews to make sure the approach makes sense.

### CODE REVIEW

File: src/dune_rules/melange/melange_rules.ml [hunk 1]
```diff
@@ -457,12 +457,27 @@ let setup_js_rules_libraries
...
+               `(implements virt_lib)`, we also set up the JS targets for the
+               modules defined in `virt_lib`.
+
+               In the cases where `virt_lib` (concrete) modules depend on any
+               virtual modules (i.e. programming against the interface), we
+               need to make sure that the JS rules that dune emits for
+               `virt_lib` depend on `concrete_lib`, such that Melange can find
+               the correct `.cmj` file, which is needed to emit the correct
+               path in `import` / `require`. *)
+            lib :: requires_link
```
Comment (anmonteiro) [inline src/dune_rules/melange/melange_rules.ml:480]: the callout here is:

- the **JS targets** of the virtual lib depend on the **library artifacts** (e.g. `.cm{i,j,t}`) of the selected (concrete) implementation

File: test/blackbox-tests/test-cases/melange/virtual_lib_compilation.t/mel.ml [hunk 2]
```diff
@@ -1 +1 @@
-print_endline Virt.t
+let () = print_endline Vlib_impl.hello
```
Comment (anmonteiro) [inline test/blackbox-tests/test-cases/melange/virtual_lib_compilation.t/mel.ml:1]: just changed this to be the same as the ml implementation

File: test/blackbox-tests/test-cases/melange/virtual_lib_compilation.t/run.t [hunk 3]
```diff
@@ -26,11 +26,16 @@ call should be emitted
   $ dune build @mel --display=short 2>&1 | grep -v 'node_modules/melange'
           melc vlib/.vlib.objs/melange/virt.{cmi,cmti}
       ocamldep impl_melange/.impl_melange.objs/virt.impl.d
-          melc vlib/.vlib.objs/melange/vlib_impl.{cmi,cmj,cmt} (exit 2)
-  File "vlib/vlib_impl.ml", line 1:
-  Error: Virt not found, it means either the module does not exist or it is a namespace
+          melc vlib/.vlib.objs/melange/vlib_impl.{cmi,cmj,cmt}
```
Comment (jchavarri) [inline test/blackbox-tests/test-cases/melange/virtual_lib_compilation.t/run.t:29]: I think the description in line 22 needs to be updated.

**Review by jchavarri — APPROVED** [review]
The fix and explanation make sense to me, the dependency flow (impl lib depends on virtual lib artifacts) seems "natural".

I wonder if there are cases where the newly added dependency can lead to cycles?

**Comment by anmonteiro** [comment]
> The fix and explanation make sense to me, the dependency flow (impl lib depends on virtual lib artifacts) seems "natural".

It’s actually the other way around: virtual lib JS depends on concrete lib artifacts (because it might call it)

**Comment by anmonteiro** [comment]
this is now ready for review. Not sure if @rgrinberg needs to take a look?

here's a summary of the included commits:

225564d5632f5c7d0af30ef8f41d71d6114ee91a fixes the issue and promotes the corresponding blackbox-test

7f90d9fa6f0765a0dc83c67e9a36b0646bbaeb10 adds the development version of Melange on 4.14 that includes a needed fix

a48d0f82537869f3a57d54c9c6d95b6fe830e5a1 promotes an unrelated test due to an upstream change in melange JS output

da681484449462bf07e3608842f748810664233c adds a changelog entry

**Review by rgrinberg — APPROVED** [review]

**Pull request merged**